### ISSUE TITLE
camelcase ignoreDestructuring fails when using destructured variables

### ISSUE THREAD

**Issue by Dentxinho** [eslint/eslint#11939]
**Tell us about your environment**

* **ESLint Version:** 6.0.1
* **Node Version:** 10.16.0
* **npm Version:** 6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
/* eslint camelcase: [error, { ignoreDestructuring: true, properties: "never" }] */
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
/* eslint camelcase: [error, { ignoreDestructuring: true, properties: "never" }] */

const { a_parameter } = query

/* These should pass */
const myVar = a_parameter
console.log(`parameter is ${a_parameter}`)
(() => a_parameter)()

/* These pass */
console.log(a_parameter)
new MyClass(a_parameter)

const myObj = { a_property: "123" }
```
The issue is reproducible on ESLint demo:

https://eslint.org/demo/#eyJ0ZXh0IjoiLyogZXNsaW50IGNhbWVsY2FzZTogW2Vycm9yLCB7IGlnbm9yZURlc3RydWN0dXJpbmc6IHRydWUsIHByb3BlcnRpZXM6IFwibmV2ZXJcIiB9XSAqL1xyXG5cclxuY29uc3QgeyBhX3BhcmFtZXRlciB9ID0gcXVlcnlcclxuXHJcbi8qIFRoaXMgc2hvdWxkIHBhc3MgKi9cclxuY29uc3QgbXlWYXIgPSBhX3BhcmFtZXRlclxyXG5jb25zb2xlLmxvZyhgcGFyYW1ldGVyIGlzICR7YV9wYXJhbWV0ZXJ9YClcclxuKCgpID0+IGFfcGFyYW1ldGVyKSgpXHJcblxyXG5cclxuY29uc29sZS5sb2coYV9wYXJhbWV0ZXIpXHJcbm5ldyBNeUNsYXNzKGFfcGFyYW1ldGVyKVxyXG5cclxuXHJcblxyXG5jb25zdCBteU9iaiA9IHsgYV9wcm9wZXJ0eTogXCIxMjNcIiB9XHJcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==

**What did you expect to happen?**
No negatives for the statements that use the variables created from object destructuring.

**What actually happened? Please include the actual, raw output from ESLint.**
No negatives when using variables as parameter to functions, but negatives when using it on string templates, variable attribution or as return value.

**Are you willing to submit a pull request to fix this bug?**
No

This looks like a related issue: #2254

**Comment by Dentxinho** [comment]
Please let me know if you need further information or examples.

**Comment by kaicataldo** [comment]
Reopening this so we can verify if this is a bug or not.

**Comment by mdjermanovic** [comment]
From #9807 it looks like the option is designed for these use cases only:

1. To avoid useless 'renaming' when one wants just to skip a property:

```js
const { some_ignored_property, ...rest} = remoteObject;
// do something with rest, nothing with some_ignored_property
```

2. To avoid 'double renaming' in situations like this:

```js
const { author_id } = this.props;
someFunc({ author_id });
// do nothing else with author_id
```

For all other use, the rule is still expected to enforce camelcase:

```js
const { a_parameter: aParameter } = query

const myVar = aParameter 
console.log(`parameter is ${aParameter}`)
(() => aParameter )()
```

**Comment by nikulis** [comment]
I just ran into this same issue; maybe the use of the rule is unclear. I use the rule because I want to prevent the **creation of** snake_case variables, but allow exceptions for destructured variables for the same reasons as listed above and in the original `ignoreDestructuring` proposal: to avoid double-renaming, and when working with data whose shape is outside the scope of my project when I don't want to rename tons of variables that could otherwise be destructured.

It seems where the confusion comes up is: once those allowed destructured snake_case variables have been created/declared, I don't want them to throw camelcase-related errors anywhere else, regardless of how they are used. As @Dentxinho originally reported,

> **What did you expect to happen?**
> No negatives for the statements that use the variables created from object destructuring.

It seems that the current rule only allows their use in function calls? Only after re-reading the rule docs a couple times am I picking up on this; it still isn't very clear to me.

---

On a related note, I'm surprised that the following is an allowed use (from the docs):

```js
/*eslint camelcase: ["error", {ignoreDestructuring: true}]*/

var { category_id: category_id } = query; // passes
```

If I were to rename a destructured variable anyways, that would negate the motivation for allowing snake_case on destructured in the first place. I would expect the behavior to instead be,

```js
/*eslint camelcase: ["error", {ignoreDestructuring: true}]*/

var { category_id } = query; // pass

var { category_id = 1 } = query; // pass

var { category_id: categoryId } = query; // pass

var { category_id: category_id } = query; // fail
```

I'm interested to hear others' thoughts, and would be happy to submit rule change proposals for either/both of the above.

**Comment by mdjermanovic** [comment]
> I just ran into this same issue; maybe the use of the rule is unclear. I use the rule because I want to prevent the **creation of** snake_case variables, but allow exceptions for destructured variables for the same reasons as listed above and in the original `ignoreDestructuring` proposal: to avoid double-renaming, and when working with data whose shape is outside the scope of my project when I don't want to rename tons of variables that could otherwise be destructured.

>It seems where the confusion comes up is: once those allowed destructured snake_case variables have been created/declared, I don't want them to throw camelcase-related errors anywhere else, regardless of how they are used. As @Dentxinho originally reported

I think that the documentation for `ignoreDestructuring` should be improved to clarify the behavior.

As you said, the rule in general prevents the creation of snake_case variables. `ignoreDestructuring` allows creation, but only if that would be for limited use (only use cases from #9807).  After reading #9807 and PR #10373 I believe this isn't a bug, though I'm not 100% sure (maybe @platinumazure can help).

> 
> It seems where the confusion comes up is: once those allowed destructured snake_case variables have been created/declared, I don't want them to throw camelcase-related errors anywhere else, regardless of how they are used.

If this isn't a bug, it might be better to make a separate "rule change request" issue.

> It seems that the current rule only allows their use in function calls? Only after re-reading the rule docs a couple times am I picking up on this; it still isn't very clear to me. 

Function/method calls are allowed as those might be globals from external libraries. It could be a mistake that the rule also allows underscores in arguments.

> On a related note, I'm surprised that the following is an allowed use (from the docs):
> 
> ```js
> /*eslint camelcase: ["error", {ignoreDestructuring: true}]*/
> 
> var { category_id: category_id } = query; // passes
> ```
> 

That's intentional to avoid overlapping with `no-useless-rename`. In general, two rules should never report essentially same error for the same code. The responsibility of the `camelcase` rule is to prevent the creation of `category_id`, which is allowed in this case because of the option, so the rule would basically report just a useless rename, which is the responsibility of `no-useless-rename`.
